### DISCUSSION TITLE
about docker off-screen mode (runing docker on server without display)

### DISCUSSION THREAD

**Discussion by gengtsh** [carla-simulator/carla#4885] [General]
hi,
i try to run carla in a server (running with ubuntu 20.04, without a display, nvidia Quadro P6000, NVIDIA-SMI 495.44 Driver Version: 495.44 CUDA Version: 11.0 ) from local windows 10 MobaXterm. 
i install docker and nvidia container toolkit, and x11vnc server on ubuntu, vnc reviewer on windows 10.

1. in the vnc reviewer (on my windows), i run make launch,  then i can run some of .py in PythonAPI/example, and it works fine. 
 
2. when i run docker version follow the instruction in both  remote ubuntu server (through ssh) and  windows 10 (vnc reviewer), i get the same output:  

    sudo docker run --privileged --gpus all --net=host -v /tmp/.X11-unix:/tmp/.X11-unix:rw carlasim/carla:0.9.12 /bin/bash ./CarlaUE4.sh -RenderOffScreen

   output:
   sh: 1:xdg-user-dir: not found
   Failed to establish dbus connectionFailed to establish dbus connectionFailed to establish dbus connection

   i also check my status dbus.service status, using sudo systemctl status dbus.service, it looks fine, i list the part of output:
   ● dbus.service - D-Bus System Message Bus
   Loaded: loaded (/lib/systemd/system/dbus.service; static; vendor preset: enabled) 
   Active: active (running) since Wed 2021-11-10 15:02:00 PST; 1 day 8h ago
   TriggeredBy: ● dbus.socket 
   Docs: man:dbus-daemon(1)
   Main PID: 1121 (dbus-daemon)
  Tasks: 1 (limit: 154144)
  Memory: 11.3M
  CGroup: /system.slice/dbus.service
  └─1121 /usr/bin/dbus-daemon --system --address=systemd: --nofork --nopidfile --systemd-activation ->
        dbus-daemon[1121]: [system] Activating via systemd: service name='org.freedesktop.nm_disp>
        dbus-daemon[1121]: [system] Successfully activated service 'org.freedesktop.nm_dispatcher'

    but, when i run docker ps, i found image carlasim/carla:0.9.12 command "/bin/bash ./CarlaUE4.sh" are created 10   minutes ago. please see the output:
    CONTAINER ID IMAGE COMMAND CREATED STATUS PORTS NAMES
    788eaaf20294 carlasim/carla:0.9.12 "/bin/bash ./CarlaUE…" 11 seconds ago Up 10 seconds blissful_yalow

    since there is a docker image running, so i try python3 dynamic_weather.py, i get some outputs, is it correct output or not?
    WARNING: Version mismatch detected: You are trying to connect to a simulator that might be incompatible with this API
    WARNING: Client API version     = 0.9.12-1-gbb4b95e0
    WARNING: Simulator API version  = 0.9.12
    Sun(alt: 11.49, azm: 314.58) Storm(clouds=65%, rain=25%, wind=40%)

    question: does this output correct or not?

3.  in the vnc reviewer client (windows 10), i run : 
    sudo docker run --privileged --gpus all --net=host -e DISPLAY=$DISPLAY carlasim/carla:0.9.12 /bin/bash ./CarlaUE4.sh
    output: 
   sh: 1:xdg-user-dir: not found
   error: XDG_RUNTIME_DIR not set in the environment
   error: XDG_RUNTIME_DIR not set in the environment
   error: XDG_RUNTIME_DIR not set in the environment

     i run echo $XDG_RUNTIME_DIR, get: /run/usr/$UID
     in this case, i check docker ps, there is no docker image running. i think there is something wrong.

     question: does this with display version suitable for the x11vnc client ?

can someone help me out? where should run docker run ? in remote ubuntu (through ssh) or in vnc reviewer client?
i would be very appreciate if someone can provide all necessary package list and provide a simple tutorial (example) to show how to run carla with docker on server without screen? thanks in advance.

**Comment by carlagx** [comment]
i have a similar problem with ubuntu 18 vm (via ssh). I solved the error with "xdg-user-dir: not found" this way:
install xdg-user-dirs xdg-utils inside the docker
https://answers.unrealengine.com/questions/1039260/view.html
still my script can not connect to the server
I start the docker with this command:
`docker run --privileged --rm -t -i --net=host --gpus all -e DISPLAY=$DISPLAY -v /tmp/.X11-unix:/tmp/.X11-unix:rw carla-server-draft bash ./CarlaUE4.sh -RenderOffScreen`

"sudo" is not necessary i think, because docker is member off sudo group? Is it right?

With Carla 9.11 and opengl everything worked fine also locally on my Ubuntu 20, however, since I switched to 9.12 I can not run Carla off screen.
Any ideas what is missing?

**Reply by DeathlyMade** [reply]
Can you tell how to run carla 9.11 in a docker container? I am running into issues

**Reply by sassy2711** [reply]
ma chuda

**Reply by carlagx** [reply]
Sorry, I don't use this version for a long time. I think I have only add `./CarlaUE4.sh -opengl` to the run docker command
https://carla.readthedocs.io/en/0.9.11/build_docker/
